storage/innobase/os/os0file.cc

```diff
@@ -675,30 +681,58 @@ os_file_create_subdirs_if_needed(
 @param[in]	request	The IO context and type
 @return the number of bytes read/written or negative value on error */
 ssize_t
-SyncFileIO::execute(const IORequest& request)
+SyncFileIO::execute_low(const IORequest& request, ssize_t n)
 {
 	ssize_t	n_bytes;
-	ut_ad(m_n > 0);
-	ut_ad(size_t(m_n) <= os_file_request_size_max);
+	ut_ad(n > 0);
+	ut_ad(size_t(n) <= os_file_request_size_max);
 
 	if (request.is_read()) {
 #ifdef _WIN32
-		n_bytes = tpool::pread(m_fh, m_buf, m_n, m_offset);
+		n_bytes = tpool::pread(m_fh, m_buf, n, m_offset);
 #else
-		n_bytes = pread(m_fh, m_buf, m_n, m_offset);
+		n_bytes = pread(m_fh, m_buf, n, m_offset);
 #endif
 	} else {
 		ut_ad(request.is_write());
 #ifdef _WIN32
-		n_bytes = tpool::pwrite(m_fh, m_buf, m_n, m_offset);
+		n_bytes = tpool::pwrite(m_fh, m_buf, n, m_offset);
 #else
-		n_bytes = pwrite(m_fh, m_buf, m_n, m_offset);
+		n_bytes = pwrite(m_fh, m_buf, n, m_offset);
 #endif
 	}
 
 	return(n_bytes);
 }
 
+/** Do the read/write
+@param[in]	request	The IO context and type
+@return the number of bytes read/written or negative value on error */
+ssize_t
+SyncFileIO::execute(const IORequest& request)
+{
+	ssize_t	n_bytes= 0;
+	ut_ad(m_n > 0);
+
+	while (size_t(m_n) > os_file_request_size_max) {
+		ssize_t n_partial_bytes= execute_low(
+			request, os_file_request_size_max);
+		if (n_partial_bytes < 0) return n_partial_bytes;
+		n_bytes+= n_partial_bytes;
+		if (n_partial_bytes != os_file_request_size_max) {
+			return n_bytes;
+		}
+		advance(os_file_request_size_max);
+	}
+
+	if (ssize_t n= execute_low(request, m_n)) {
+		if (n < 0) return n;
+		n_bytes += n;
+	}
+
+	return(n_bytes);
+}
```

Comment:
Let’s avoid the old InnoDB style formatting in new code, and let’s only document the parameters in the function declaration in the header file:
```c++
ssize_t SyncFileIO::execute(const IORequest &request)
{
  ssize_t n_bytes= 0;
  ut_ad(m_n > 0);
/// ...
  return n_bytes;
}
```